### ISSUE TITLE
twisted/web/xmlrpc support for <nil/> types.

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#7188]
|<img alt="virus's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">| virus reported|
|-|-|
|Trac ID|trac#469|
|Type|enhancement|
|Created|2004-01-07 06:41:10Z|

Attachments:

* [xmlrpc_allowNone.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/e3e/e3e097dc79d8161b2a2448f6c0930a8b081cd013/54c3497f30be6612f50cdde8addde7ee195acec3.patch) (2537 bytes) - added by virus on 2004-01-07 06:41:10Z - 
* [xmlrpc_allowNone_by_xmlrpclibVersion.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/e3e/e3e097dc79d8161b2a2448f6c0930a8b081cd013/762aae3ff6373acf3b4bcb46a7b5528c9551e4b7.patch) (2587 bytes) - added by virus on 2004-01-07 23:40:25Z - 
* [xmlrpc_allowNone_by_xmlrpclibVersion_rev2.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/e3e/e3e097dc79d8161b2a2448f6c0930a8b081cd013/8bc1c8cdd64f405cc86354046973a952b9a27bfa.patch) (2593 bytes) - added by virus on 2004-01-07 23:50:24Z - 
* [xmlrpc_allowNone_by_xmlrpclibVersion_rev3.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/e3e/e3e097dc79d8161b2a2448f6c0930a8b081cd013/70d5047aa2b66e7a8a6c8a07e86956286aca930e.patch) (2411 bytes) - added by virus on 2004-01-08 00:22:32Z - 
* [xmlrpc_allowNone_by_xmlrpclibVersion_rev4.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/e3e/e3e097dc79d8161b2a2448f6c0930a8b081cd013/eb8e78a12694735b02a4e8a84f31a82595268b71.patch) (2417 bytes) - added by virus on 2004-01-08 03:06:53Z - 
* [xmlrpc.diff](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/e3e/e3e097dc79d8161b2a2448f6c0930a8b081cd013/688714645378a27a96ffb76213eb38d9141c5301.diff) (1161 bytes) - added by spiv on 2006-06-26 06:33:18Z - Updated patch
* [xmlrpc_469.diff](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/e3e/e3e097dc79d8161b2a2448f6c0930a8b081cd013/c15790730a37dba8bd7b8cfacc41d59f828f1a6e.diff) (3008 bytes) - added by therve on 2006-06-27 08:35:11Z - 
* [text_xmlrpc_469.diff](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/e3e/e3e097dc79d8161b2a2448f6c0930a8b081cd013/e4de2cc5c25904f40c7f7a8df754379af2c01606.diff) (1305 bytes) - added by therve on 2006-06-27 08:35:28Z - 

<details><summary>Searchable metadata</summary>

```
trac-id__469 469
type__enhancement enhancement
reporter__virus virus
priority__highest highest
milestone__ 
branch__ 
branch_author__ 
status__closed closed
resolution__fixed fixed
component__web web
keywords__ 
time__1073457670000000 1073457670000000
changetime__1470651271493472 1470651271493472
version__ 
owner__ 
cc__itamarst cc__virus cc__spiv cc__therve
```
</details>

**Comment by twisted-trac** [comment]
|<img alt="virus's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|virus commented|
|-|-|

```
#!html
<pre>
python-2.3 and the bundled xmlrpclib-1.0.1 provides the &lt;nil/> extension for
XML-RPC. twisted/web/xmlrpc is not aware of this extension, yet.

Follow ups:
 * http://ontosys.com/xml-rpc/extensions.html
 * http://www.python.org/doc/current/lib/module-xmlrpclib.html

TODO:
 * The documentation needs to be updated according to these changes.

Attached File:
The neccesary changes in xmlrpc.py in order to support the allow_none extension
of xmlrpclib.
</pre>
```

**Comment by twisted-trac** [comment]
|<img alt="virus's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|virus commented|
|-|-|

```
#!html
<pre>
It might be problematic to use the python version for checking if allow_none is
supported. Just think of someone using python2.2 with xmlrpclib 1.0.1 downloaded
from  http://www.pythonware.com/products/xmlrpc/

Attached File:
The neccesary changes in xmlrpc.py in order to support the allow_none extension
of xmlrpclib which use the xmlrpclib version instead of the python version.

ATTENTION:
 * Don't use the xmlrpc_allowNone_by_xmlrpclibVersion.patch but use revision 2.
I made some serious mistake in the previous revision.
</pre>
```

**Comment by twisted-trac** [comment]
|<img alt="virus's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|virus commented|
|-|-|

```
#!html
<pre>
Revision 3 of the patch uses xmlrpclib.__version__ instead of "from xmlrpclib
import __version__ as xmlrpclibVersion" since xmlrpclib is already imported at
the beginning of the file.
</pre>
```

**Comment by twisted-trac** [comment]
|<img alt="virus's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|virus commented|
|-|-|

```
#!html
<pre>
I recently realized that comparing a list and a tuple does not work as expected
so I need to release a 4th revision of the patch which replaces all (1,0,1)
tuples by [1,0,1] lists.
</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)|@itamarst commented|
|-|-|

```
#!html
<pre>
The patch fails the tests. No idea why!
</pre>
```

**Comment by twisted-trac** [comment]
|<img alt="spiv's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|spiv commented|
|-|-|

I've attached a patch against current SVN to add an `allowNone` parameter to `t.w.xmlrpc.XMLRPC`.

I haven't updated `QueryFactory.__init__`, the way it uses `*args` means that you can't just add `allowNone=False` to the function signature without breaking existing code.  You could do it via `**kwargs` I guess, but perhaps that API should be rethought rather than just hacking this on.

The existing tests still pass with this change, though obviously a test for this new feature would be a good idea...

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)|@therve commented|
|-|-|

I may be wrong, but the allowNone arguments can be passed to the QueryFactory constructor, as long as you provide a default value, no ? I attach a patch doing this and add a test case.

**Comment by twisted-trac** [comment]
|[<img alt="jyknight's avatar" src="https://avatars.githubusercontent.com/u/873903?s=50" width="50" height="50">](https://github.com/jyknight)|@jyknight commented|
|-|-|

Adding more arguments before *args _does_ change the interface to QueryFactory. It seems like that might be okay -- I'm not sure, but it looks like it might be supposed to be an internal interface? If it's changed, though, "args" should be made a normal arg, rather than a restarg.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun removed owner|
|-|-|

I've put this into a branch - xmlrpc-none-469 - with a few minor changes: the test method has been broken up into two pieces and documented somewhat; I renamed QueryFactory to _QueryFactory to make sure that if anyone was incorrectly using it, their code will break very loudly instead of silently doing the wrong thing; I also made the change to args that James suggested.

**Comment by twisted-trac** [comment]
|[<img alt="radix's avatar" src="https://avatars.githubusercontent.com/u/227068?s=50" width="50" height="50">](https://github.com/radix)|@radix set owner to @exarkun|
|-|-|

+1, this matches my own code which has used xmlrpc with allowNone. Merge it.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun set status to `closed`|
|-|-|

(In [17771]) Merge xmlrpc-none-469

Authors: virus, therve, exarkun
Reviewer: radix
Fixes [#7188](https://github.com/twisted/twisted/issues/7188)

This adds support for the </nil> xml-rpc extension type to the
twisted.web xml-rpc server and client.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|<img alt="GitHub's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_14"></a>|GitHub commented|
|-|-|

In changeset 6fafccd6ca7f8821f1a25bb4aa1801f4802833f8
```
#!CommitTicketReference repository="" revision="6fafccd6ca7f8821f1a25bb4aa1801f4802833f8"
Merge pull request #469 from twisted/8719-rodrigc-stringtransport-py3

Author: rodrigc
Reviewer: acabhishek942
Fixes: #8719

Fix twisted/test/test_stringtransport.py  on Python 3
```
